### PR TITLE
Hw 23

### PR DESCRIPTION
**Pull request by KrohIgor**

### CODE REVIEW

File: src/main/java/mate/academy/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+    @Override
+    public Order update(Order order) {
+        String query = "DELETE FROM `internet-shop`.orders_products\n"
+                + "WHERE order_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            preparedStatement.setLong(1, order.getOrderId());
+            preparedStatement.executeUpdate();
+            insertProductsToOrdersProducts(order, connection);
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:84]: what if some products where deleted from order?(not in our case, but still)

File: src/main/java/mate/academy/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        String query = "DELETE FROM `internet-shop`.shopping_carts_products\n"
+                + "WHERE shopping_cart_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            preparedStatement.setLong(1, shoppingCart.getShoppingCartId());
+            preparedStatement.executeUpdate();
+            for (Product product : shoppingCart.getProducts()) {
+                String queryUpdate = "INSERT INTO "
+                        + "`internet-shop`.shopping_carts_products values(?, ?)";
+                PreparedStatement statement = connection.prepareStatement(queryUpdate);
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: let's move statement before loop in order not to create the same statement  for each  product

File: src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        String query = "SELECT * FROM `internet-shop`.users \n"
+                + "INNER JOIN users_roles ON users.user_id = users_roles.user_id\n"
+                + "INNER JOIN roles ON users_roles.role_id = roles.role_id\n"
+                + "WHERE users.user_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            preparedStatement.setLong(1, id);
+            ResultSet resultSet = preparedStatement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(getUserFromResultSet(resultSet));
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: what if user have a few roles, would we get them also or only the role from the first row?

File: src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            throw new DataProcessingException("Couldn't get User with id - " + id, e);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public List<User> getAll() {
+        String query = "SELECT * FROM `internet-shop`.users \n"
+                + "INNER JOIN users_roles ON users.user_id = users_roles.user_id\n"
+                + "INNER JOIN roles ON users_roles.role_id = roles.role_id";
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: what is the total number of rows you will get in the result? the number of user or the number of users_roles?same problem as in previous comment

File: src/main/java/mate/academy/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 1 above)
Comment (KrohIgor) [reply src/main/java/mate/academy/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:84]: specified

File: src/main/java/mate/academy/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java (hunk 2 above)
Comment (KrohIgor) [reply src/main/java/mate/academy/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: fixed

File: src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java (hunk 3 above)
Comment (KrohIgor) [reply src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: fixed

File: src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java (hunk 4 above)
Comment (KrohIgor) [reply src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: fixed

File: src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                    connection.prepareStatement(queryUserRole);
+            for (Role role : user.getRoles()) {
+                preparedStatementRole.setString(1, String.valueOf(role.getRoleName()));
+                ResultSet resultSet = preparedStatementRole.executeQuery();
+                if (resultSet.next()) {
+                    long roleId = resultSet.getLong("role_id");
+                    preparedStatementUserRole.setLong(1, user.getUserId());
+                    preparedStatementUserRole.setLong(2, roleId);
+                    preparedStatementUserRole.executeUpdate();
+                }
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: let's make a separate method for this

**Review by dmytro-hryhoruk — APPROVED** [review]

File: src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java (hunk 5 above)
Comment (KrohIgor) [reply src/main/java/mate/academy/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: done

**Pull request merged**